### ISSUE TITLE
Kotlin Codegen: Data class constructor default values that are evaluated from other fields don't work correctly

### ISSUE THREAD

**Issue by JulleR** [square/moshi#690]
When using Moshi's Kotlin Codegen (1.6.0 and 1.7.0), the value of _endTimeMs_ is always 0, assuming the value is not included in the parsed JSON, but it was set correctly when using Kotlin reflection:
```
@JsonClass(generateAdapter = true)
data class SessionInfo constructor(
        val ref: Ref,
        val location: Location,        
        val startTime: LocalDateTime,
        val endTime: LocalDateTime? = null,
        val receivedTime: Long = System.currentTimeMillis()),
        val startTimeMs: Long = startTime.toMilliseconds(),
        val endTimeMs: Long = endTime?.toMilliseconds() ?: 0)
}
```

In my case, _endTimeMs_ is never included in the parsed JSON so this is to easy fix by moving it to the class body. But I think this case should either be fixed in Codegen, or at least have some Lint error about it because this can easily cause bugs.

_startTimeMs_ works because _startTime_ is a required field, so _startTimeMs_ gets initialized by the constructor when the first result-instance is created. But _endTime_ is not a required field, thus _endTimeMs_ becomes 0 in the first result-instance and the copy-constructor just copies that value to the second result-instance.

From the generated adapter:
```
var result = SessionInfo(
        ref = ref ?: throw JsonDataException("Required property 'ref' missing at ${reader.path}"),
        location = location ?: throw JsonDataException("Required property 'location' missing at ${reader.path}"),
        startTime = startTime ?: throw JsonDataException("Required property 'startTime' missing at ${reader.path}"))

result = result.copy(
        endTime = if (endTimeSet) endTime else result.endTime,
        receivedTime = receivedTime ?: result.receivedTime,
        startTimeMs = startTimeMs ?: result.startTimeMs,
        endTimeMs = endTimeMs ?: result.endTimeMs)
```

**Comment by JulleR** [comment]
Probably just documenting this behavior as a known issue for codegen is enough. At least then users know about it.

**Comment by ZacSweers** [comment]
This is fixed on master via #896 and will be available in the next release
